subsys: power: clean up function names & update documentation

mnkp opened the pull request:
This PR attempts to clean up naming of functions and Kconfig options used by the power management subsystem. The detailed changes are described in the commit messages.

I've kept the `sys_pm_` prefix for the power management subsystem functions. It's not really the best choice. It's neither meaningful nor consistent with naming of Kconfig options but modifying it would require matching changes involving device power management functions. That's much more work but maybe we should do it. A better prefix would be e.g. `sys_power_`.

To hide functions which should not be visible to the user I prefixed them with a leading underscore which I believe is its intended usage.

The documentation of the power management subsystem is updated and aligned with the implementation.

mnkp committed:
power: rename Low Power States to Sleep States

There exists SoCs, e.g. STM32L4, where one of the low power modes
reduces CPU frequency and supply voltage but does not stop the CPU. Such
power modes are currently not supported by Zephyr.

To facilitate adding support for such class of power modes in the future
and to ensure the naming convention makes it clear that the currently
supported power modes stop the CPU this commit renames Low Power States
to Slep States and updates the documentation.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
power: clean up system power managment function names

This commit cleans up names of system power management functions by
assuring that:
- all functions start with 'sys_pm_' prefix
- API functions which should not be exposed to the user start with '_'
- name of the function hints at its purpose

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

dbkinder reviewed (dismissed):
Looks OK to me...

ulfalizer commented on arch/Kconfig:
```diff
@@ -349,25 +349,25 @@ config ARCH_HAS_THREAD_ABORT
 # individual SoC.
 #
-config HAS_STATE_LOW_POWER_1
+config HAS_STATE_SLEEP_1
```
Wondering if `HAS_SLEEP_STATE_1` would work. I like it when it reads naturally in English.

mnkp replied:
I was wondering about that myself. The reason I kept `HAS_STATE_` pattern was to make sure the names match definitions of `enum power_states` in `include/power.h`. The enum defines constants such as `SYS_POWER_STATE_SLEEP_1`, `SYS_POWER_STATE_DEEP_SLEEP_3`. They - in turn - use `SYS_POWER_STATE_` to match the enum name.

To use `HAS_SLEEP_STATE_1` I believe we should rename also the enum identifiers. It's a bigger change but yes, we could to it.

ulfalizer replied:
Sounds good to me at least.

mnkp committed:
power: modify HAS_STATE_SLEEP_ Kconfig options

Add SYS_POWER_ prefix to HAS_STATE_SLEEP_, HAS_STATE_DEEP_SLEEP_
options to align them with names of power states they control.
Following is a detailed list of string replacements used:
s/HAS_STATE_SLEEP_(\d)/HAS_SYS_POWER_STATE_SLEEP_$1/
s/HAS_STATE_DEEP_SLEEP_(\d)/HAS_SYS_POWER_STATE_DEEP_SLEEP_$1/

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp committed:
power: rename residency policy Kconfig options

Rename power managment subsystem Kconfig options describing minimum
residency to make them easier to identify with respective policy.

Following is a detailed list of string replacements used:
s/SYS_PM_SLEEP_(\d)_MIN_RES/SYS_PM_MIN_RESIDENCY_SLEEP_$1/
s/SYS_PM_DEEP_SLEEP_(\d)_MIN_RES/SYS_PM_MIN_RESIDENCY_DEEP_SLEEP_$1/

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

mnkp commented:
I've renamed `HAS_STATE_SLEEP_1` to `HAS_SLEEP_STATE_1` as requested by @ulfalizer. This is a larger change and if deemed to obtrusive I can remove it. I decided not to change names of the constants defined by `enum power_states` as some of them didn't work that well. E.g. renaming `SYS_POWER_STATE_AUTO` to `SYS_POWER_AUTO_STATE` didn't seem to be a good choice.

I've added one more commit cleaning up Kconfig options related to residency policy.

@benpicco changes in this PR will influence #13494. Please review.

mnkp committed:
doc: update power management subystem documentation

This commit brings power management subystem documentation up to date
with the implementation. The main changes since the last time the
documentation was updated include introduction of power managment
policies.

Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

pizi-nordic commented on arch/Kconfig:
```diff
@@ -351,42 +351,42 @@ config ARCH_HAS_THREAD_ABORT
 # individual SoC.
 #
-config HAS_STATE_LOW_POWER_1
+config HAS_SLEEP_STATE_1
```
HAS_SLEEP_STATE_X matches with SYS_POWER_STATE_SLEEP_X.
Could we have the same order of "STATE" and "SLEEP" in both?

pizi-nordic replied:
This looks good. Then I would suggest HAS_SYS_POWER_{DEEP_}SLEEP_STATE_X as a matching option.

pizi-nordic commented on doc/reference/power_management/index.rst:
```diff
@@ -90,109 +89,97 @@ is a thread ready to run or if an external event occurred.
 
-Suspend Hook function
-=====================
+It is an application responsibility to set up a wake up event. A wake up event
```
This will probably change, as the kernel have to track time and system timer is usually stopped on low power modes.

pizi-nordic commented on doc/reference/power_management/index.rst:
```diff
@@ -90,109 +89,97 @@ is a thread ready to run or if an external event occurred.
 
-Suspend Hook function
-=====================
+It is an application responsibility to set up a wake up event. A wake up event
```
Could you please tell here how the app should hook to the PM in order to setup the wake-up event.

mnkp replied:
The problem of setting a wake up event was largely ignored when designing the power management subsystem. This part we still need to discuss and agree upon. Nevertheless, as far as I can tell looking at the current source code, the paragraph holds true.

mnkp commented:
recheck

mnkp commented:
I had another look at the power states naming scheme and decided to take a bit different approach. Rather than changing `HAS_STATE_SLEEP_1` to `HAS_SLEEP_STATE_1` and along it all literals defined by `enum power_states` I renamed `HAS_STATE_SLEEP_1` to `HAS_SYS_POWER_STATE_SLEEP_1`.

This way the literals read well in English, have their own namespace and we avoid inconsistency in naming of `enum power_states`.

If that's not OK I'll change to what we've agreed upon.

pizi-nordic commented:
It is OK. I just wanted clear match between enum and Kconfig.

galak commented:
@ramakrishnapallala can you review this.

mnkp commented:
@dbkinder could you review the last commit "doc: update power management subystem documentation". You've reviewed the previous commits already and they haven't changed.

dbkinder reviewed (dismissed):
Thanks for the update.  Looks like I may have missed some items in the previous review.

mnkp commented:
@dbkinder Thanks for the review! I've updated the PR.
